examples/pytorch/distributed_stl10/stl10_example.py

```diff
@@ -0,0 +1,360 @@
+    n_gpu_per_node = world_size // n_nodes
+
+    # define master address and master port
+    hostnames = subprocess.check_output(['scontrol', 'show', 'hostnames', os.environ['SLURM_JOB_NODELIST']])
+    master_addr = hostnames.split()[0].decode('utf-8')
+
+    # set environment variables for 'env://'
+    os.environ['MASTER_ADDR'] = master_addr
+    os.environ['MASTER_PORT'] = str(29500)
+    os.environ['WORLD_SIZE'] = str(world_size)
+    os.environ['RANK'] = str(global_rank)
+
+    # define whether this is the master process / if we are in distributed mode
+    is_master = node_id == 0 and local_rank == 0
+    multi_node = n_nodes > 1
+    multi_gpu = world_size > 1
+
+    # summary
+    PREFIX = "%i - " % global_rank
+    print(PREFIX + "Number of nodes: %i" % n_nodes)
+    print(PREFIX + "Node ID        : %i" % node_id)
+    print(PREFIX + "Local rank     : %i" % local_rank)
+    print(PREFIX + "Global rank    : %i" % global_rank)
+    print(PREFIX + "World size     : %i" % world_size)
+    print(PREFIX + "GPUs per node  : %i" % n_gpu_per_node)
+    print(PREFIX + "Master         : %s" % str(is_master))
+    print(PREFIX + "Multi-node     : %s" % str(multi_node))
+    print(PREFIX + "Multi-GPU      : %s" % str(multi_gpu))
+    print(PREFIX + "Hostname       : %s" % socket.gethostname())
+
+    # set GPU device
+    torch.cuda.set_device(local_rank)
+
+    print("Initializing PyTorch distributed ...")
+    torch.distributed.init_process_group(
+        init_method='env://',
+        backend='nccl',
+    )
+
+    print("Initialize Model...")
+    # Construct Model
+    model = models.resnet18(pretrained=False, num_classes=10).cuda()
+    # Make model DistributedDataParallel
+    model = DistributedDataParallel(model, device_ids=[local_rank], output_device=local_rank)
+
+    # define loss function (criterion) and optimizer
+    criterion = nn.CrossEntropyLoss().cuda()
+    optimizer = SGD(model.parameters(), learning_rate, momentum=0.9, weight_decay=1e-4)
+
+    print("Initialize Dataloaders...")
+    transform = transforms.Compose(
+        [transforms.Resize(224),
+         transforms.ToTensor(),
+         transforms.Normalize(mean=[0.485, 0.456, 0.406],
+                              std=[0.229, 0.224, 0.225])])
+
+    # Initialize Datasets.
+    trainset = datasets.STL10(root=os.environ["SCRATCH"], split='train', download=False, transform=transform)
+    valset = datasets.STL10(root=os.environ["SCRATCH"], split='test', download=False, transform=transform)
+
+    # Create DistributedSampler to handle distributing the dataset across nodes
+    # This can only be called after torch.distributed.init_process_group is called
+    train_sampler = DistributedSampler(trainset)
+    val_sampler = DistributedSampler(valset)
+
+    # Create the Dataloaders to feed data to the training and validation steps
+    train_loader = DataLoader(trainset,
+                              batch_size=batch_size,
+                              num_workers=10,
```

Comment:
I haven't run this specific example but in my code, I have observed some performance drop when using a number of workers equal to the number of cpus assigned to the task. I think having too many dataloading processes can hurt inter-node communication.

Regarding perforamnce, I have also observed that adding more interfaces can help, e.g. `export NCCL_SOCKET_IFNAME=ib0,bond1,bond0,eno1,eno2` but I am not sure if this would help for this simple MNIST example.

Reply:
>I haven't run this specific example but in my code, I have observed some performance drop when using a number of workers equal to the number of cpus assigned to the task. I think having too many dataloading processes can hurt inter-node communication.

Indeed, `num_workers` adds some overhead to data loading that might not be worth it if the transform pipeline is simple and the dataset is small (which is the case here) but I think its a good baseline to start with.

> Regarding perforamnce, I have also observed that adding more interfaces can help, e.g. export NCCL_SOCKET_IFNAME=ib0,bond1,bond0,eno1,eno2 but I am not sure if this would help for this simple MNIST example.

I didn't know about this. Maybe we should add it to the config?

Reply:
Hmm, I have recently performed some tests on a different task and I couldn't see a difference between using these variables or letting NCCL find the correct network interfaces by itself. Maybe the performance difference I noticed at the time came from something else. Therefore, I don't think it it necessary to add it in this PR, especially that it can cause errors if the specified interfaces are not available on some nodes (not all are available on all queues).